tests/integration/cocoon/test_cocoon_holdings.py

```diff
@@ -1110,3 +1113,114 @@ def test_return_unmatched_holding_handles_empty_holdings_exception_in_lusid(self
+                "data/holdings-example-duplicate-shk.csv",
+                ["Security Description"],
+                "DuplicateSubHoldingKeysProvided",
+                False,
+            ],
+            [
+                "Failed with non-existent portfolio",
+                "data/holdings-example-single-holding.csv",
+                ["Security Description"],
+                "PortfolioNotFound",
+                True,
+            ],
+        ]
+    )
+    def test_failed_load_from_data_frame_holdings_returns_useful_errors_and_skips_validation_logic(
+        self, _, file_name, sub_holding_keys, error_name, skip_portfolio,
+    ) -> None:
+        """
+        Test that a failed holding upload is handled gracefully by the whole load_from_data_frame function.
+        If any upload causes an error the validation logic will not run and all the errors will be returned
+        by load_from_data_frame.
+
+        :param str file_name: The name of the test data file
+        :param list[str] sub_holding_keys: The column names for any sub-holding-keys
+        :param str error_name: The name of the ApiException that is hit by this error
+        :param bool skip_portfolio: Flag whether to skip the creation of a test portfolio
+
+        :return: None
+        """
+        # Unchanged vars that have no need to be passed via param (they would count as duplicate lines)
+        scope = "unmatched_holdings_test"
+        mapping_required = {
+            "code": "FundCode",
+            "effective_at": "Effective Date",
+            "tax_lots.units": "Quantity",
+        }
+        mapping_optional = {"tax_lots.cost.currency": "Local Currency Code"}
+        identifier_mapping = {
+            "Isin": "ISIN Security Identifier",
+            "Sedol": "SEDOL Security Identifier",
+            "Currency": "is_cash_with_currency",
+            "ClientInternal": "Client Internal",
+        }
+        property_columns = ["Prime Broker"]
+        properties_scope = "operations001"
+        sub_holding_key_scope = None
+        return_unmatched_items = True
+        holdings_adjustment_only = False
+
+        data_frame = pd.read_csv(Path(__file__).parent.joinpath(file_name))
+
+        if not skip_portfolio:
+            # Create the portfolios
+            portfolio_response = cocoon.cocoon.load_from_data_frame(
+                api_factory=self.api_factory,
+                scope=scope,
+                data_frame=data_frame,
+                mapping_required={
+                    "code": "FundCode",
+                    "display_name": "FundCode",
+                    "base_currency": "Local Currency Code",
+                },
+                file_type="portfolio",
+                mapping_optional={"created": "Created Date"},
+            )
+
+            # Assert that all portfolios were created without any issues
+            self.assertEqual(len(portfolio_response.get("portfolios").get("errors")), 0)
```

Comment:
In order to be truly independent, we would have different portfolios for each parametrised test case. Alternatively, we could initialise and destroy the portfolio within the setup() and teardown() functions.

Reply:
The portfolios are unique for each test case, insofar as the portfolio is created based on the portfolio present in the test data, which is unique and provided by the params. The reason for the skipping portfolio logic here is so that test cases which do not want a portfolio to exist, are not forced to create one. If we want to go the way of setup and teardown functions we could, but it may get a bit complicated due to the fact that the portfolio(s) are different for each test, especially with other tests in this class that are not in this MR.